### PR TITLE
[Lite] Remove full-size FP32 temporaries from the FSDP2 optimizer step

### PR DESCRIPTION
**Pull request by ISEEKYAN**
# [Lite] Remove full-size FP32 temporaries from the FSDP2 optimizer step

## Motivation

FSDP2 OOMs more easily than it should. Two places in the optimizer step allocate
transient FP32 buffers proportional to the *largest* gradient, at exactly the
point where memory is already at its high-water mark.

**1. Grad norm.** `grad_clip.py` computed `t.to(float32).pow(2).sum()` per grad.
This keeps **two** full-size FP32 temporaries alive at once — the cast result is
still referenced while `pow()` allocates its own output — i.e. **4x the bytes of
a BF16 grad**. On a large fused expert weight that is hundreds of MB of pure
transient. It also launches 3 kernels per parameter in a Python loop.

**2. AdamW step.** `FP32AdamW._prepare_grad()` did `grad.detach().to(float32)`
for every parameter — another 2x the BF16 grad, in the *same* phase. Fixing only
the grad norm would leave the end-to-end peak unchanged.

mcore does not have either problem: `get_grad_norm_fp32` uses TE/apex's fused
`multi_tensor_l2norm`, and its grads are already FP32 `main_grad` buffers by the
time they reach the optimizer, so no cast is needed at all.

## Changes

**Grad norm → TE fused `multi_tensor_l2norm`** (same kernel as mcore
`megatron/core/optimizer/clip_grads.py`): FP32 accumulation inside the kernel,
zero temporaries, one launch per bucket instead of one per parameter.

- `grad_clip.py`: new `fused_sq_sum()`, bucketed by `(dtype, device)` —
  `multi_tensor_applier` dispatches on the tensor list's scalar type, so
  mixed-dtype lists are not safe to pass. mcore never hit this because its grads
  are uniformly FP32; lite mixes BF16 and FP32.
- Non-contiguous tensors deliberately stay off the multi-tensor path (they would
  need a full `contiguous()` copy — exactly the allocation being removed) and use
  `torch.linalg.vector_norm`, which is likewise temporary-free. CPU tensors use
  the same path, which keeps the CPU-only unit tests working.
- `Partial`-placement groups still process one grad at a time: `_local_grad()`
  calls `full_tensor()` there, and collecting the group into a list first would
  keep several global-shaped tensors alive simultaneously.
- `_group_local_abs_max()` had the identical double-temporary defect in
  `.to(dtype).abs().max()`; now `vector_norm(inf)`. This also fixes the
  comparison silently being done in BF16 for BF16 grads.
- `adamw.py`: `local_grad_sq_sum()` shares the fused path, so plain-tensor grads
  do not fall back to a full-size allocation.
- Removed the unused `chunk_size_numel` plumbing (no callers in tree).

**AdamW step:** `_prepare_grad()` no longer materializes an FP32 copy.
`exp_avg`/`exp_avg_sq` are already FP32 and `add_()`/`addcmul_()` promote a BF16
operand to the accumulator dtype inside the kernel, so the copy bought nothing.
For `cpu_update`, the D2H transfer now moves the grad in its own dtype, halving
it as well.

## Numerics

Grad norm: `norm` then square instead of a direct squared sum. Relative error is
~1e-7 (and it is *less* prone to FP32 overflow, not more). All new tests assert
against a float64 reference.

AdamW: **bitwise identical**. Verified over 4 steps across `param`, `exp_avg`,
`exp_avg_sq` and `master_param`, for both the GPU path and `cpu_update`.

## Tests

Added to `test_fsdp2_unit.py`:

- `fused_sq_sum` vs a float64 reference on mixed BF16/FP32/empty inputs
- one multi-tensor launch **per dtype**, not per parameter (guards the bucketing)
- non-contiguous tensors never reach the multi-tensor kernel
- BF16 grads accumulate in FP32 on both the sq-sum and inf-norm paths
- `FP32AdamW` bitwise equivalence vs the old FP32-copy path, parametrized over
  `cpu_update`
- **CUDA-only**: peak allocation during the norm must stay under 1 MB for a
  64M-element BF16 grad (the old code allocated 512 MB there). Numerical tests
  cannot catch a memory regression — this is the one that guards the actual point
  of the change.

`_is_fused_eligible()` exists as a monkeypatch seam so the CPU-only unit tests can
exercise the bucket path without a GPU.

## Not in this PR

`denom = exp_avg_sq.sqrt()` in the AdamW step is still a full-size FP32
temporary. Removing it needs a fused Adam kernel. `maybe_build_te_fused_adam_optimizer()`
already wires up TE `FusedAdam(master_weights=True)` behind
`fsdp2_use_te_fused_adam` (default off) — flipping that default is the better
lever, but it changes optimizer state format and needs convergence validation, so
it belongs in its own PR.

### CODE REVIEW

**Comment by ISEEKYAN** [comment]
Combined verdict: **one-line blocker, then good to go.** Two independent passes — a review panel over the diff, and the repo's own gate actually executed on 8xH100 with `base` and this branch in the same container.

## BLOCKER (one line): the new CUDA test has no `gpus` marker

`test_fsdp2_unit.py:683`

```python
@pytest.mark.skipif(not torch.cuda.is_available(), reason="TE multi_tensor_l2norm is CUDA-only.")
def test_fused_sq_sum_allocates_no_full_size_temporary():
```

`_test_harness/markers.py` states that an absent `gpus` marker means a CPU test, so this lands in the `gpus=0` cpu suite, CUDA is hidden, and it skips. `pytest_worker.py` counts any skip as a failure, and `runner.py` stops at the first non-PASS suite. Measured, standard profile, no arguments, 8 GPUs allocated:

```
base    (06d046fe3): cpu suite  701 passed / 0 failed / 5 skipped
patched (42061d07e): cpu suite  706 passed / 0 failed / 6 skipped   <- the 6th is this test
```

So the branch adds one skip of its own, and the test that guards this PR's entire claim never runs. The sibling test at `:397` in the same file shows the intended pattern (`@pytest.mark.gpus(1)` above the `skipif`). Fix:

```python
@pytest.mark.gpus(1)
@pytest.mark.skipif(not torch.cuda.is_available(), reason="TE multi_tensor_l2norm is CUDA-only.")
```

Related: the five pre-existing skips on `base` come from `test_hf_weights_gpu.py` missing the same marker, which is what #178 fixes. Until #178 lands, the cpu suite fails and no GPU suite runs at all — so this PR's GPU test cannot execute even once it is marked. #186 effectively depends on #178 for its own coverage.

## With the marker fixed: zero regression, and the claim verifies

Both arms carrying the #178 fix plus the marker, same container, same job:

```
                                   base            patched
suite=cpu                          701 passed/0sk  706 passed/0sk
suite=1gpu-csa-thd-cp              3               3
suite=1gpu-fsdp2-offload-gpu       2               2
suite=1gpu-fsdp2-unit              1               2     <- the new CUDA test runs and passes
suite=1gpu-glm5-lite-static        3 + 4           3 + 4
suite=1gpu-hf-weights-gpu          3               3
suite=1gpu-hf-weights-streaming    1               1
suite=1gpu-qwen-lite-forward-smoke 2               2
suite=1gpu-router-aux-loss-gate    2               2
suite=2gpu-distopt-checkpoint      1               1
suite=2gpu-fsdp2-offload-ckpt      3               3
suite=2gpu-glm5-lite-cp-smoke      2 passed/3 failed (identical on both arms, pre-existing on main)
```

The `< 1 MB` peak-allocation assertion holds against the real TE kernel, not a mock. Everything else is byte-for-byte the same between arms; the only red suite is red on unpatched `main` too.

## Panel caveats (eight reviewers, no blockers)

1. `sharded_grad_sq_sum` dropping `chunk_size_numel` was flagged as an API break. I would downgrade it: there is not a single caller in the tree, and on `main` the parameter is only ever passed by the function to itself. It was dead configuration and removing it is a cleanup.
2. `adamw.py:14` importing `fused_sq_sum` at module level binds the `state.py:11` import path to TE. mlite seat called this major, contract seat minor.
3. `_prepare_grad` now relies on `add_`/`addcmul_` promoting into the FP32 accumulators. That is an implicit contract held up by one bitwise test at a single seed and shape. Panel split major/minor; the `torch.equal` test is what kept it at minor.
4. Uneven-shard DTensor paths (`copy_local_tensor_to_param_`, `_scale_grad_`'s `mul_`) rest on comments, with no DTensor-level regression test. The evidence seat kept this at major.
5. Coverage gaps: partial-placement grad norm is untested; `fused_sq_sum` is compared at `rel=1e-5`; "fused path is CUDA + contiguous only" is not stated in the docstring.

One thing worth noting about method: the panel read the diff closely enough to pair `adamw.py:232` with its test at `:732-733`, and still did not catch the marker problem — that is only visible by running the gate. Conversely the DTensor and implicit-promotion concerns above would not surface from running it. Neither pass substitutes for the other.

Panel run `review-moe-619257d2`, verdict approve-with-caveats.
